google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/SchemaCompact.java

```diff
@@ -76,24 +102,473 @@ private TableId getTableId(String tableName) {
   }
 
   /**
-   * Checks if the userSchema is compatible with the table's current schema for writing. The current
-   * implementatoin is not complete. If the check failed, the write couldn't succeed.
+   * Checks if proto schema is supported.
    *
-   * @param tableName The name of the table to write to.
-   * @param userSchema The schema user uses to append data.
-   * @throws IllegalArgumentException the check failed.
+   * @param protoSchema
+   * @return True if protoSchema is supported
+   * @throws IllegalArgumentException if schema is invalid
    */
-  public void check(String tableName, Descriptors.Descriptor userSchema)
+  public static void isSupported(Descriptors.Descriptor protoSchema)
       throws IllegalArgumentException {
-    Table table = bigquery.getTable(getTableId(tableName));
-    Schema schema = table.getDefinition().getSchema();
-    // TODO: We only have very limited check here. More checks to be added.
-    if (schema.getFields().size() != userSchema.getFields().size()) {
+    HashSet<Descriptors.Descriptor> allMessageTypes = new HashSet<>();
+    allMessageTypes.add(protoSchema);
+    String protoSchemaName = protoSchema.getName();
+    if (!isSupportedImpl(protoSchema, allMessageTypes, protoSchemaName)) {
       throw new IllegalArgumentException(
-          "User schema doesn't have expected field number with BigQuery table schema, expected: "
-              + schema.getFields().size()
-              + " actual: "
-              + userSchema.getFields().size());
+          "Proto schema "
+              + protoSchemaName
+              + " is not supported: contains nested messages of more than 15 levels.");
     }
   }
+
+  /**
+   * Actual implementation that checks if a proto schema is supported.
+   *
+   * @param protoSchema
+   * @param allMessageTypes Keeps track of all message types seen, and make sure they don't repeat
+   *     as recursive protos are not supported.
+   * @param protoScope Debugging purposes to show error if messages are nested.
+   * @return True if field type and option is supported
+   * @throws IllegalArgumentException if schema is invalid
+   */
+  private static boolean isSupportedImpl(
+      Descriptors.Descriptor protoSchema,
+      HashSet<Descriptors.Descriptor> allMessageTypes,
+      String protoScope)
+      throws IllegalArgumentException {
+
+    for (Descriptors.FieldDescriptor field : protoSchema.getFields()) {
+      String currentProtoScope = protoScope + "." + field.getName();
+      if (!isSupportedType(field)) {
```

Comment:
If the user field belongs to unknown fields, and it has unsupported type, we should allow it. That's why I think you need to build the isSupported into schema check.

Reply:
You can add a unit test for this case.

Reply:
Merge isSupported and isCompatible, and added a test that checks if proto schema is accepted when it has a unsupported field, but this field is an unknown field that is allowed by the boolean flag allowUnknownFields.